### PR TITLE
Improve shared memory protection

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following:**

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/FTL/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:** 

## 10

---

**Store and check shared memory ownership before resizing shared memory objects.**

The scenario we are resolving here is the following: FTL is already running and someone(/-thing) deletes all the shared memory objects. If another instance of FTL is then started, it doesn't know a previous process is running and creates new SHM objects for itself. Both processes can - in theory - run just fine without touching each other as they are both actually using *different* shared memory objects. You are right when you think this is bizarre. The reason is that the first one has the shared memory objects still mapped into memory, i.e., "deleting" them from disk only removes the visible file handles. The next FTL instance creates files with the very same name, however, they are also new and distinct files, i.e., their memory will point elsewhere. Both instances can now run in parallel just fine *until* one of them needs to resize a shared memory objects. If one of the shared memory events now gets resized to a size larger than it was before BUT smaller than what the other FTL instance is expecting, the other instance will instantaneously crash with a `SIGSEGV` (a bus error to be precise).

This PR resolves this by storing the PID of the SHM object creator in the settings object.

Each FTL instance reloads this shared memory instance now before performing any potentially dangerous operation and checks if the shared memory files on disk are still owned by this process. If this is not the case, we are in serious trouble and exit immediately. This should allow the second instance (you could call it the "rightful owner" of the current existing SHM objects) a fairly good chance to never even notice this and continue to operate just fine.

Resolves the root causes behind #1463, #1448 and possibly other similar issues in the past by preventing a certain "mis"configuration to have the bad effect it had for these users.

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -55,27 +55,31 @@ int main (int argc, char* argv[])
 	logg("########## FTL started on %s! ##########", hostname());
 	log_FTL_version(false);
 
+	// Process pihole-FTL.conf
```
Comment (yubiuser) [inline src/main.c:58]: Why did you move reading the conf file before initializing the shared memory?

Comment (DL6ER) [reply src/main.c:58]: This has been mentioned in the commit message of https://github.com/pi-hole/FTL/pull/1464/commits/fc303aeba52cd2e742ea6cc2b62e3d6c8a5f13ac. Sorry, I should have included this in the PR description:

> Read FTL config *before* initializing shared memory to ensure we are aware of a possible DEBUG_SHMEM flag when creating the shared memory segments

Without this, any `DEBUG_SHMEM` code was useless in `init_shmem()`.

**Review by yubiuser — APPROVED** [review]
